node/network/collator-protocol/src/collator_side.rs

```diff
@@ -0,0 +1,1009 @@
+	messages::{
+		AllMessages, CollatorProtocolMessage, RuntimeApiMessage, RuntimeApiRequest,
+		NetworkBridgeMessage,
+	},
+};
+use polkadot_node_network_protocol::{
+	v1 as protocol_v1, View, PeerId, PeerSet, NetworkBridgeEvent, RequestId,
+};
+use polkadot_node_subsystem_util::{
+	request_validators_ctx,
+	request_validator_groups_ctx,
+};
+
+#[derive(Default)]
+struct State {
+	/// Our id.
+	our_id: CollatorId,
+
+	/// The para this collator is collating on.
+	/// Starts as `None` and is updated with every `CollateOn` message.
+	collating_on: Option<ParaId>,
+
+	/// Track all active peers and their views
+	/// to determine what is relevant to them.
+	peer_views: HashMap<PeerId, View>,
+
+	/// Our own view.
+	view: View,
+
+	/// Posessed collations.
+	///
+	/// We will keep up to one local collation per relay-parent.
+	collations: HashMap<Hash, (CandidateReceipt, PoV)>,
+
+	/// Our validator groups active leafs. 
+	our_validators_groups: HashMap<Hash, Vec<ValidatorId>>,
+
+	/// Validators we know about via `ConnectToValidators` message.
+	/// 
+	/// These are the only validators we are interested in talking to and as such
+	/// all actions from peers not in this map will be ignored.
+	/// Entries in this map will be cleared as validator groups in `our_validator_groups`
+	/// go out of scope with their respective deactivated leafs.
+	known_validators: HashMap<PeerId, ValidatorId>,
+}
+
+/// Distribute a collation.
+/// 
+/// Figure out the core our para is assigned to and the relevant validators.
+/// Issue a connection request to these validators.
+/// If the para is not scheduled or next up on any core, at the relay-parent,
+/// or the relay-parent isn't in the active-leaves set, we ignore the message
+/// as it must be invalid in that case - although this indicates a logic error
+/// elsewhere in the node.
+async fn distribute_collation<Context>(
+	ctx: &mut Context,
+	state: &mut State,
+	id: ParaId,
+	receipt: CandidateReceipt,
+	pov: PoV,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = CollatorProtocolMessage>
+{
+	let relay_parent = receipt.descriptor.relay_parent;
+
+	// This collation is not in the active-leaves set.
+	if !state.view.contains(&relay_parent) {
+		warn!(
+			target: TARGET,
+			"Distribute collation message parent {:?} is outside of our view",
+			relay_parent,
+		);
+
+		return Ok(());
+	}
+
+	// We have already seen collation for this relay parent.
+	if state.collations.contains_key(&relay_parent) {
+		return Ok(());
+	}
+
+	// Determine which core the para collated-on is assigned to.
+	// If it is not scheduled then ignore the message.
+	let our_core = match determine_core(ctx, id, relay_parent).await? {
+	    Some(core) => core,
+	    None => {
+			warn!(
+				target: TARGET,
+				"Looks like no core is assigned to {:?} at {:?}", id, relay_parent,
+			);
+			return Ok(());
+		}
+	};
+
+	// Determine the group on that core and
+	// TODO: the next group on that core.
```

Comment:
This should be resolvable given current information from the runtime API. 

`fn validator_groups` gives a `Vec<Vec<ValidatorIndex>>` and `GroupRotationInfo`. I'm assuming we've gotten the core index from `group_rotation_info.group_for_core`.

So then it's just

```rust
let next_group_index = GroupIndex((current_group_index.0 + 1) % n_groups);
```

Find the validators of the group the same way and connect to them the same way.

Note that this does cause an error at the end of sessions, where you're prospectively connecting to a "next" group which never happens, but that's fine because 99% of the time this works.